### DISCUSSION TITLE
Alternative emulator support

### DISCUSSION THREAD

**Discussion by grahambates** [prb28/vscode-amiga-assembly#306] [General]
As it appears that FS-UAE is effectively dead, I've been looking at alternatives as I don't think I'll be able to keep up with patching it long term. We're already on shakey ground as the 4.x development version was in an incomplete state before we started backporting changes.

The good news is that [AmiBerry](https://github.com/BlitterStudio/amiberry) appears to be a very strong alternative. When I looked at this previously it was targeted at low spec platforms and lacking some key features, but since the 6.x version it seems to have good feature parity with WinUAE. More importantly is very actively developed.

I've had a go at porting the debug server code to this, and I already have a working version. There are a few annoyances around things like mouse capture, but nothing we can't fix.

In the longer term [Quaesar](https://github.com/theblacklotus/quaesar) looks really promising too.

It would be nice to revisit integrating with vAmiga too, as I know a fair bit of progress was made on this.

**Comment by prb28** [comment]
Didn't know for fs-uae ! I thought it was a healthy project. That not good news.

The best would be a good alternative that work for all platforms to reduce the development and maintenance time.

A long time ago vAmiga was first in my list and as you said some work was done. I don't know how the project is going. There was a wasm version that was promising as a multi platform alternative.

**Reply by prb28** [reply]
Quaesar and Amiberry look good, and if they share the code base with winuae the port and maintenance for the debugger could be not too difficult as you experienced.

**Comment by grahambates** [comment]
Yeah, Frode seems to have completely gone missing and hasn't committed anything in nearly two years. This is indeed very concerning and I hope he's ok. I had been backporting WinUAE changes myself, but this is getting increasing difficult as the code bases diverge. This is more of an issue for Bartman's debugger, because it uses WinUAE internal data structures for profiling and these need to be kept in sync.

I think the best option is to add AmiBerry as an additional config option for now, with a view to making it default and dropping FS-UAE in the longer term.

There's a strong argument to continue to use WinUAE on Windows as it's kind of the gold standard in terms of features and compatibility, but AmiBerry supports all other platforms and is very close. In fact it's much more similar to WinUAE than FS-UAE was in terms of command line options, config files etc, so you can pretty much treat them like-for-like without too many special cases.

I also agree that vAmigaWeb offers some very interesting possibilities. [This](https://www.youtube.com/watch?v=dpDK6S2CTHQ) is a great example of the sort of thing that can be done, and it offers the possibility of something much more integrated with vscode, using web views. It is currently OCS 68000 only though, so we'd still need to maintain support for other emulators.
